src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +159,236 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
+ *
+ * @param [in] _locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array) \
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/** Release the ptr_array lock
+ *
+ * @param [in] _locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array) \
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieves a value from the locked array.
+ *
+ * @param [in]  locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in]  element_index      Index to retrieve the value from.
+ * @param [out] var                Filled with the value.
+ *
+ * @return Whether the value is present and valid.
+ *
+ * Complexity: O(1)
+ */
+static inline int
+ucs_ptr_array_locked_lookup(ucs_ptr_array_locked_t *locked_ptr_array,
+                            unsigned element_index, void **var)
+{
+    int present;
+
+    ucs_ptr_array_locked_acquire_lock(locked_ptr_array);
+    present = ucs_ptr_array_lookup(&locked_ptr_array->super, element_index,
+                                   *var);
+    ucs_ptr_array_locked_release_lock(locked_ptr_array);
+
+    return present;
+}
+
+
+/**
+ * Get the current size of the locked ptr array
+ *
+ * @param [in] locked_ptr_array      Pointer to a locked ptr array.
+ *
+ * @return Size of the locked ptr array.
+ */
+static inline unsigned
+ucs_ptr_array_locked_get_size(ucs_ptr_array_locked_t *locked_ptr_array)
+{
+    ucs_ptr_array_t *ptr_array = &locked_ptr_array->super;
+
+    return ucs_ptr_array_get_size(ptr_array);
+}
 
 
+/**
+ * If foreach locked ptr_array is finalized, releases lock.
+ *
+ * @param [in] locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in] element_index      The current for loop index.
+ *
+ * @return is_continue_loop for the for() loop end condition.
+ */
+static inline int
+__ucx_ptr_array_locked_foreach_finalize(ucs_ptr_array_locked_t *locked_ptr_array,
+                                        uint32_t element_index)
+{
+    int is_continue_loop =
+        (element_index < locked_ptr_array->super.size);
+
+    /* Release lock at the end */
+    if (!is_continue_loop) {
+        ucs_ptr_array_locked_release_lock(locked_ptr_array);
+    }
+
+    return is_continue_loop;
+}
+
+
+/**
+ * Iterate over all valid elements in the locked array.
+ *
+ * Please notice that using break or return are not allowed in 
+ * this implementation.
+ * Using break or return would require releasing the lock before by calling,
+ * ucs_ptr_array_locked_release_lock(_locked_ptr_array);
+ *
+ * @param [out] _var                Pointer to current array element in the foreach.
+ * @param [out] _index              Index variable to use as iterator (unsigned).
+ * @param [in]  _locked_ptr_array   Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_for_each(_var, _index, _locked_ptr_array) \
+    ucs_ptr_array_locked_acquire_lock(_locked_ptr_array); \
```

Comment:
better do it inside for initialization phase, so the macro will generate a single statement

Reply:
Fixed.